server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -658,68 +673,257 @@ protected void doRun() {
+                            }
+
+                            int i = 0;
+                            for (InferenceResults inferenceResults : results.transformToLegacyFormat()) {
+                                String fieldName = inferenceFieldNames.get(i++);
+                                @SuppressWarnings("unchecked")
+                                Map<String, Object> inferenceFieldMap = (Map<String, Object>) rootInferenceFieldMap.computeIfAbsent(
+                                    fieldName,
+                                    k -> new HashMap<String, Object>()
+                                );
+
+                                inferenceFieldMap.put(INFERENCE_FIELD, inferenceResults.asMap("output").get("output"));
+                                inferenceFieldMap.put(TEXT_FIELD, docMap.get(fieldName));
                             }
-                            responses.set(bulkItemResponse.getItemId(), bulkItemResponse);
+
+                            docRef.close();
                         }
-                        maybeFinishHim();
-                    }
 
-                    @Override
-                    public void onFailure(Exception e) {
-                        // create failures for all relevant requests
-                        for (BulkItemRequest request : requests) {
+                        @Override
+                        public void onFailure(Exception e) {
+
                             final String indexName = request.index();
                             DocWriteRequest<?> docWriteRequest = request.request();
-                            BulkItemResponse.Failure failure = new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e);
+                            BulkItemResponse.Failure failure = new BulkItemResponse.Failure(
+                                indexName,
+                                docWriteRequest.id(),
+                                new IllegalArgumentException("Error performing inference: " + e.getMessage(), e)
+                            );
                             responses.set(request.id(), BulkItemResponse.failure(request.id(), docWriteRequest.opType(), failure));
+                            // make sure the request gets never processed again
+                            bulkShardRequest.items()[request.id()] = null;
+
+                            docRef.close();
                         }
-                        maybeFinishHim();
+                    });
+                }
+            }
+        }
+
+        private static List<String> getFieldNamesForInference(
+            Map.Entry<String, Set<String>> fieldModelsEntrySet,
+            Map<String, Object> docMap
+        ) {
+            List<String> inferenceFieldNames = new ArrayList<>();
+            for (String inferenceField : fieldModelsEntrySet.getValue()) {
+                Object fieldValue = docMap.get(inferenceField);
+
+                // Perform inference on string, non-null values
```

Comment:
We'll want to handle arrays too?

Reply:
This is tricky, do we separate out the arrays here or have the inference API return an array of responses?